File: FluidStudio/FluidStudio/Physics/Fluid/EmitterScene.cs
using FluidStudio.VDB;
using PG.Control.OpenGL;
using PG.Core.Math;
using PG.Scene;
using System.Linq;
using CreateLabels = PG.EmitterSceneCreateLabels;
using UpdateLabels = PG.EmitterSceneUpdateLabels;


namespace FluidStudio.Physics.Fluid
{
    public class EmitterScene : IPhysicsScene
    {
        public int Id { get; private set; }

        private int SourceParticleSystemId { get; set; }

        public string ParticleFilePath { get; private set; }

        public float ParticleRadius { get; set; } = 0.5f;

        public int VolumeId { get; private set; }

        public string Name { get; set; }

        public float Density { get; set; } = 1000.0f;

        public float Stiffness { get; set; } = 1.0f;

        public Vector3d InitialVelocity { get; set; } = new Vector3d(0, 0, 0);

        public float Viscosity { get; set; } = 1.0f;

        public int StartTimeStep { get; set; } = 0;

        public int EndTimeStep { get; set; } = 100;

        public int Interval { get; set; } = 10;

        public FluidFileExportModel ExportModel { get; }
            = new FluidFileExportModel();

        public IPhysicsScene Parent { get; private set; }

        private readonly SceneList scenes;

        public EmitterScene(SceneList scenes, SolverScene parent, VDBModel vdb, Canvas3d canvas)
        {
            this.scenes = scenes;
            this.Parent = parent;
            var command = new PG.CLI.PhysicsCommand(CreateLabels.CommandNameLabel);
            command.Execute(scenes.Adapter);
            this.Id = command.GetResult<int>(CreateLabels.NewIdLabel);
            this.VolumeId = vdb.CreateVDBVolume("Volume", false);
            canvas.BuildShader(scenes, VolumeId);
        }

        public void SetParticlesFromFile(VDBModel vdb, FileIOModel ioModel, Canvas3d canvas, string particleFilePath, double particleRadius)
        {
            var ext = System.IO.Path.GetExtension(particleFilePath);
            if (ext != ".vdb")
            {
                return;
            }
            var absolutePath = ioModel.ToFullPath(particleFilePath);
            // TODO 前のPSのクリア処理．
            var ids = vdb.Read(absolutePath, (float)particleRadius);
            if (ids.Count() > 0)
            {
                this.SourceParticleSystemId = ids[0];
                this.ParticleFilePath = particleFilePath;
            }
        }

        public void ExportFiles(VDBModel vdb, int timeStep)
        {
            //ExportModel.ExportFiles(scenes, this, vdb, timeStep);
        }

        public void ConvertPSToVolume(VDBModel vdb, double radius)
        {
            vdb.ConvertPSToVolume(this.Id, this.VolumeId, radius);
        }

        public void Send()
        {
            var command = new PG.CLI.PhysicsCommand(UpdateLabels.CommandNameLabel);
            command.SetArg(UpdateLabels.IdLabel, Id);
            command.SetArg(UpdateLabels.ParticleSystemIdLabel, SourceParticleSystemId);
            command.SetArg(UpdateLabels.ParticleRadiusLabel, ParticleRadius);
            command.SetArg(UpdateLabels.DensityLabel, Density);
            command.SetArg(UpdateLabels.StiffnessLabel, Stiffness);
            command.SetArg(UpdateLabels.ViscosityLabel, Viscosity);
            command.SetArg(UpdateLabels.InitialVelocityLabel, InitialVelocity);
            //command.SetArg(UpdateLabels.IsBoundary, IsBoundary);
            command.SetArg(UpdateLabels.StartStepLabel, StartTimeStep);
            command.SetArg(UpdateLabels.EndStepLabel, EndTimeStep);
            command.SetArg(UpdateLabels.IntervalLabel, Interval);
            command.SetArg(UpdateLabels.NameLabel, Name);
            command.Execute(scenes.Adapter);
        }

        public void Delete()
        {
            scenes.Delete(SourceParticleSystemId);
            scenes.Delete(Id);
            scenes.Delete(VolumeId);
        }

    }
}

File: Blender/CrystalPython/physics/emitter_scene.py
from physics.physics_labels import EmitterSceneCreateCommand, EmitterSceneUpdateCommand
from scene.scene import Scene
from CrystalPLI import *
from scene.scene_labels import ParticleSystemGetCommand

class EmitterScene :
    def __init__(self, scene) :
        self.__scene = scene
        self.__id = 0
        self.source_particle_system_id = 0
        self.particle_radius = 1.0
        self.density = 1.0
        self.stiffness = 0.25
        self.viscosity = 10.0
        self.__start_step = 0
        self.__end_step = 100
        self.__interval = 5
        self.__initial_velocity = Vector3df(0,0,0)

    def get_id(self) :
        return self.__id

    def set_start_step(self, start) :
        self.__start_step = start

    def set_end_step(self, end) :
        self.__end_step = end

    def set_interval(self, interval) :
        self.__interval = interval

    def set_initial_velocity(self, v) :
        self.__initial_velocity = Vector3df(v[0],v[1],v[2])

    def create(self) :
        create_physics_command(EmitterSceneCreateCommand.CommandNameLabel)
        execute_command(self.__scene.world)
        self.__id = get_result_int(EmitterSceneCreateCommand.NewIdLabel)

    def send(self) :
        create_physics_command(EmitterSceneUpdateCommand.CommandNameLabel)
        set_arg_int(EmitterSceneUpdateCommand.IdLabel, self.__id)
        set_arg_int(EmitterSceneUpdateCommand.ParticleSystemIdLabel, self.source_particle_system_id)
        set_arg_float(EmitterSceneUpdateCommand.ParticleRadiusLabel, self.particle_radius)
        set_arg_float(EmitterSceneUpdateCommand.DensityLabel, self.density)
        set_arg_float(EmitterSceneUpdateCommand.StiffnessLabel, self.stiffness)
        set_arg_float(EmitterSceneUpdateCommand.ViscosityLabel, self.viscosity)
        set_arg_int(EmitterSceneUpdateCommand.StartStepLabel, self.__start_step)
        set_arg_int(EmitterSceneUpdateCommand.EndStepLabel, self.__end_step)
        set_arg_int(EmitterSceneUpdateCommand.IntervalLabel, self.__interval)
        set_arg_vector3df(EmitterSceneUpdateCommand.InitialVelocityLabel, self.__initial_velocity)
        is_ok = execute_command(self.__scene.world)
        return is_ok

    def get_positions(self) :
        create_scene_command(ParticleSystemGetCommand.CommandNameLabel)
        set_arg_int(ParticleSystemGetCommand.PSIdLabel, self.__id)
        is_ok = execute_command(self.__scene.world)
        return get_result_vector3dd_vector(ParticleSystemGetCommand.PositionsLabel)
